ambari-server/src/main/java/org/apache/ambari/server/api/handlers/ReadHandler.java

```diff
@@ -121,4 +125,17 @@ private void addFieldsToQuery(Request request, Query query) {
       query.addProperty(propertyId, entry.getValue());
     }
   }
+
+  /**
+   * FORMAT predicate is ignored, but if it is format=summary, it should be added to queryProperty in Read Request
+   *
+   * @param request the current request
+   * @param query   the associated query
+   */
+  private void addSummaryQueryToQuery(Request request, Query query) {
+    String uriString = request.getURI();
+    if (StringUtils.isNotBlank(uriString) && uriString.indexOf("format=summary") != -1) {
+      query.addLocalProperty(HostResourceProvider.HOST_SUMMARY_PROPERTY_ID);
+    }
+  }
```

Comment:
This seems wrong to add to a generic handler like the ReadHandler. Why is this even needed if you're using a renderer to do the work?

Reply:
Now based on previous discussion, I have removed all summary resourceprovider and summary render. Also the "format" in uri is filtered and is not regarded as a predicate, so there is no way for  readhandler to know what exact property that this read request wants to get. "summary" is a special read  request and in current uri parsing, it is actually ignored. If we compare with alert summary(that is the only implementation of summary in Ambari), it uses a separate resourceprovider and renderer to realize it. Originally I did the same way, but I got some comments to suggest not create new resourceprovider and/or new renderer in consideration of mantainenance and possibly duplicated code.  So I think if the uri inlcudes " format=summary" queryParam, it means we should do summary query instead of individual property query. This should be one of a generic operation.

Reply:
That's right, it's not regarded as a predicate. However, the `getRenderer()` method of the `ResourceDefinition` can determine which renderer to use. The `BaseRequest` has the handle to the renderer name already.

```
  /**
   * Parse the query string for the {@link QueryLexer#QUERY_FORMAT} property and obtain
   * a renderer from the associated resource definition based on this property value.
   */
  private void parseRenderer() {
    String rendererName = isMinimal() ? "minimal" :
        m_uriInfo.getQueryParameters().getFirst(QueryLexer.QUERY_FORMAT);
    m_renderer = m_resource.getResourceDefinition().
        getRenderer(rendererName);
  }
```

Reply:
The resourceProvider does not know the renderer, it can only know this through  resource.type. We can do in another way: Create a new resource.type=HostSummary, and create a hostsummaryRenderer. So hostResourceProvider can handle Request based on resource.type differently. But this implementation will introudce new resource.type and new resourcerenderer. Moreover, the hostservice will create different resourceInstances based on if there has queryParam and queryParam is "format=summary", either "resource.type=Host" or "resource.type=HostSummary". If you think this is ok, I can modify my code to implement it.

Reply:
That is correct - the resource provider doesn't know the renderer. It satisfies the predicate to provide the concrete resources. From the resources, the correct renderer is applied to the dataset to return the data in a new way. 

I'm not understanding why this can't follow the same pattern as other renderers in the system. The data is post-processed by the renderer in order to provide a summary of data instead of the individual resources themselves.

Reply:
I agree with @jonathan-hurley .  This is exactly why renderers exist.  Now it may not be as clean as a summary DTO out of the db, but it surely can be done.